File: Q1-Q50/Q20.Valid Parentheses/Solution.java
// Solution to the Q20 "Valid Parentheses"
// https://leetcode.com/problems/valid-parentheses/description/

class Solution {
    private boolean isPair(char open, char close) {
        if (open == '(' && close == ')') {
            return true;
        }
        else if (open == '{' && close == '}') {
            return true;
        }
        else if (open == '[' && close == ']') {
            return true;
        }
        return false;
    }

    public boolean isValid(String s) {
        if (s == null || s.isEmpty()) {
            return true;
        }
        List<Character> stack = new ArrayList<Character>();
        char[] chs = s.toCharArray();
        for (char c : chs) {
            if (c == '{' || c == '(' || c == '[') {
                stack.add(c);
            }
            else {
                if (stack.isEmpty()) {
                    return false;
                }
                char open = stack.remove(stack.size() - 1);
                if (!isPair(open, c)) {
                    return false;
                }
            }
        }
        return stack.isEmpty();
    }
}

File: Q1-Q50/Q20.Valid Parentheses/Solution.js
// JS solution to Q20. "Valid Parentheses"
// Reference: https://leetcode.com/problems/valid-parentheses/description/

/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
    let stack = [];
    // to use array as stack:
    // push: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/push
    // pop: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/pop

    // iterate through each character of the string
    for (const c of s) {
        switch (c) {
            // if it is a open-character, add it to the stack
            case '(':
                stack.push('(');
                break;
            case '[':
                stack.push('[');
                break;
            case '{':
                stack.push('{');
                break;
            // NOTE: when the array is empty, the pop will return "undefined",
            // so it will also immediately return false
            case ']':
                if (stack.pop() !== '[') {
                    return false;
                }
                break;
            case ')':
                if (stack.pop() !== '(') {
                    return false;
                }
                break;
            case '}':
                if (stack.pop() !== '{') {
                    return false;
                }
                break;
            default:
                break;
        }
    }
    return stack.length === 0;
};
